File: src/ObjectManager/Object.Tes/FilePacks/Records/040-ACRE.Placed creature.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class ACRERecord : Record
    {
        public override string ToString() => $"GMST: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public FMIDField<Record> NAME; // Base
        public REFRRecord.DATAField DATA; // Position/Rotation
        public List<CELLRecord.XOWNGroup> XOWNs; // Ownership (optional)
        public REFRRecord.XESPField? XESP; // Enable Parent (optional)
        public FLTVField? XSCL; // Scale (optional)
        public BYTVField? XRGD; // Ragdoll Data (optional)

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "NAME": NAME = new FMIDField<Record>(r, dataSize); return true;
                case "DATA": DATA = new REFRRecord.DATAField(r, dataSize); return true;
                case "XOWN": if (XOWNs == null) XOWNs = new List<CELLRecord.XOWNGroup>(); XOWNs.Add(new CELLRecord.XOWNGroup { XOWN = new FMIDField<Record>(r, dataSize) }); return true;
                case "XRNK": ArrayUtils.Last(XOWNs).XRNK = new IN32Field(r, dataSize); return true;
                case "XGLB": ArrayUtils.Last(XOWNs).XGLB = new FMIDField<Record>(r, dataSize); return true;
                case "XESP": XESP = new REFRRecord.XESPField(r, dataSize); return true;
                case "XSCL": XSCL = new FLTVField(r, dataSize); return true;
                case "XRGD": XRGD = new BYTVField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/040-ACRE.Placed creature.swift
//
//  ACRERecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class ACRERecord: Record {
    public override var description: String { return "GMST: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var NAME: FMIDField<Record>! // Base
    public var DATA: REFRRecord.DATAField! // Position/Rotation
    public var XOWNs: [CELLRecord.XOWNGroup]? // Ownership (optional)
    public var XESP: REFRRecord.XESPField? // Enable Parent (optional)
    public var XSCL: FLTVField? // Scale (optional)
    public var XRGD: BYTVField? // Ragdoll Data (optional)
   
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "NAME": NAME = FMIDField<Record>(r, dataSize)
        case "DATA": DATA = REFRRecord.DATAField(r, dataSize)
        case "XOWN":
            if XOWNs == nil { XOWNs = [CELLRecord.XOWNGroup]() }; XOWNs!.append(CELLRecord.XOWNGroup(XOWN: FMIDField<Record>(r, dataSize)))
        case "XRNK": XOWNs!.last!.XRNK = r.readT(dataSize)
        case "XGLB": XOWNs!.last!.XGLB = FMIDField<Record>(r, dataSize)
        case "XESP": XESP = REFRRecord.XESPField(r, dataSize)
        case "XSCL": XSCL = r.readT(dataSize)
        case "XRGD": XRGD = r.readBYTV(dataSize)
        default: return false
        }
        return true
    }
}
